### PR TITLE
Fixed #24727 -- Prevented ClearableFileInput from masking exceptions on Python 2

### PR DESCRIPTION
**Pull request by bluezio**
Please see the following Trac ticket:

https://code.djangoproject.com/ticket/24727#ticket

### CODE REVIEW

**Comment by timgraham** [comment]
Could you please add a test to demonstrate the usefulness of the change?

**Comment by timgraham** [comment]
Also, it would be nice to structure the code in such a way that we can easily revert to the current version when we drop Python 2 compatibility. Using `six.PY2`, for example.

**Comment by bluezio** [comment]
The problem with six.PY2 is that hasattr wasn't changed until Python 3.2: sites using Python 3.0 and 3.1 would be affected again by this issue if the code were reverted to the current version. Would this be acceptable?

I'll add a test for the change.

**Comment by timgraham** [comment]
Yes, it's okay because Django 1.9 supports Python 2.7 and 3.3+.

**Comment by bluezio** [comment]
I've added a test case and changed the code so it uses six.PY2. Thanks for the suggestions :-).

**Comment by timgraham** [comment]
buildbot, add to whitelist.

File: django/forms/widgets.py [hunk 1]
```diff
@@ -377,7 +377,14 @@ def is_initial(self, value):
         """
         Return whether value is considered to be initial value.
         """
-        return bool(value and hasattr(value, 'url'))
+        if six.PY2:
```
Comment (timgraham) [inline django/forms/widgets.py]: Could you add a comment `# hasattr() masks exceptions on Python 2`

File: django/forms/widgets.py [hunk 2]
```diff
@@ -377,7 +377,14 @@ def is_initial(self, value):
...
+            try:
+                getattr(value, 'url')
+                return bool(value)
```
Comment (timgraham) [inline django/forms/widgets.py]: this could go in the else block of the try/except

File: django/forms/widgets.py [hunk 3]
```diff
@@ -377,7 +377,14 @@ def is_initial(self, value):
...
+            except AttributeError:
+                return False
+        else:
```
Comment (timgraham) [inline django/forms/widgets.py]: don't need an else

**Comment by bluezio** [comment]
I've revised the patch to take into account your comments. Thanks for the cleanup :-).

**Comment by timgraham** [comment]
merged in 5c412dd8a724b263489c1bd7a2fea381460665d7, thanks!

**Pull request closed without merging**